### ISSUE TITLE
java.lang.IllegalAccessError in lzycompute in Java 11

### ISSUE THREAD

**Issue by leitoh0** [scala/bug#11781]
Newer JVMs (11 at least) throw a java.lang.IllegalAccessError on detecting an `update to non-static final field <foo> attempted from a different method (<bar>$lzycompute) than the initializer method <init>`.

`fail$lzcompute()` in the following code would contain such an illegal `putfield`.

```scala
class C
class Fail(c: C) {
  lazy val fail = c
}
```
Perhaps the best solution is to de-(java-)finalize single-use private immutable fields, at least the ones slated for nullification?

`MixinTransformer.lazyValNullables` could exclude (java-)final fields, exposing potential GC cost and turning itself mostly pointless in the process, so that may be a less-desirable option.

It is possible that there are other places where such now-illegal field nullification could be taking place (aside from `lzycompute`). IMHO backend does not seem to do so for _fields_, though.

**Comment by leitoh0** [comment]
E.g. something of this nature:
```patch
--- a/src/compiler/scala/tools/nsc/transform/Mixin.scala
+++ b/src/compiler/scala/tools/nsc/transform/Mixin.scala
@@ -453,8 +453,10 @@ abstract class Mixin extends Transform with ast.TreeDSL with AccessorSynthesis {
             deriveDefDef(dd) {
               case blk@Block(stats, expr) =>
                 assert(dd.symbol.originalOwner.isClass, dd.symbol)
-                def nullify(sym: Symbol) =
+                def nullify(sym: Symbol) = {
+                  sym.accessedOrSelf.setFlag(MUTABLE)
                   Select(gen.mkAttributedThis(sym.enclClass), sym.accessedOrSelf) === NULL
+                }
                 val stats1 = stats ::: fieldsToNull.map(nullify)
                 treeCopy.Block(blk, stats1, expr)
               case tree =>
```

**Comment by hrhino** [comment]
Hi @leitoh0:

That looks reasonable to me. Would you mind submitting it as a patch?

**Comment by leitoh0** [comment]
Hi @hrhino,

Thanks for checking. Not at all: https://github.com/scala/scala/pull/8514
Do you want a 2.13.x PR separately (same diff), or is the above sufficient?

**Comment by hrhino** [comment]
Thanks! 2.12.x only is fine; it gets merged into 2.13.x periodically.

**Comment by som-snytt** [comment]
Requires `-target:9` and up, if anyone tries it and wonders.
